File: 04 April/04-04/equationValidation.js
function isValidEquation(equation) {
  /* Solution using eval()
  const { left, right } = /(?<left>.+)=(?<right>.+)/.exec(equation).groups;

  return eval(left) === eval(right);
  */

  // Solution performing manual evaluation
  const regex = /\d+|[-/+*=]/g;
  const equationArray = [...equation.matchAll(regex).map(match => {
    const op = Number(match[0]);
    return Number.isNaN(op) ? match[0] : op;
  })];
  
  while (["*", "/"].some(operator => equationArray.includes(operator))) {
    for (let index in equationArray) {
      index = Number(index);
      let newValue;
      
      if (equationArray[index] === "*")
        newValue = equationArray[index - 1] * equationArray[index + 1];

      if (equationArray[index] === "/")
        newValue = equationArray[index - 1] / equationArray[index + 1];
      
      if (newValue !== undefined) {
        equationArray.splice(index - 1, 3, newValue);
        break;
      }
    }
  }

  while (["+", "-"].some(operator => equationArray.includes(operator))) {
    for (let index in equationArray) {
      index = Number(index);
      let newValue;
      
      if (equationArray[index] === "+")
        newValue = equationArray[index - 1] + equationArray[index + 1];

      if (equationArray[index] === "-")
        newValue = equationArray[index - 1] - equationArray[index + 1];
      
      if (newValue !== undefined) {
        equationArray.splice(index - 1, 3, newValue);
        break;
      }
    }
  }

  return equationArray[0] === equationArray.at(-1);
}

// --- TEST SUITE ---

const testsText = String.raw`
1. isValidEquation("2 + 2 = 4") should return true.
2. isValidEquation("2 + 3 - 1 = 4") should return true.
3. isValidEquation("8 / 2 = 4") should return true.
4. isValidEquation("10 * 5 = 50") should return true.
5. isValidEquation("2 - 2 = 0") should return true.
6. isValidEquation("2 + 9 / 3 = 5") should return true.
7. isValidEquation("20 - 2 * 3 = 14") should return true.
8. isValidEquation("2 + 5 = 6") should return false.
9. isValidEquation("10 - 2 * 3 = 24") should return false.
10. isValidEquation("3 + 9 / 3 = 4") should return false.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);

File: 04 April/04-04/equation_validation.py
def is_valid_equation(equation):
    regex = r"\d+|[/+*=-]"
    equation_list = [parse_op(match[0]) for match in re.finditer(regex, equation)]

    while any(operator in equation_list for operator in ["*", "/"]):
        for index, op in enumerate(equation_list):
            new_value = None

            if equation_list[index] == "*":
                new_value = equation_list[index - 1] * equation_list[index + 1]

            if equation_list[index] == "/":
                new_value = equation_list[index - 1] / equation_list[index + 1]

            if new_value != None:
                del equation_list[index - 1:index + 2]
                equation_list.insert(index - 1, new_value)
                break

    while any(operator in equation_list for operator in ["+", "-"]):
        for index, op in enumerate(equation_list):
            new_value = None

            if equation_list[index] == "+":
                new_value = equation_list[index - 1] + equation_list[index + 1]

            if equation_list[index] == "-":
                new_value = equation_list[index - 1] - equation_list[index + 1]

            if new_value != None:
                del equation_list[index - 1:index + 2]
                equation_list.insert(index - 1, new_value)
                break

    return equation_list[0] == equation_list[-1]

def parse_op(op):
    try:
        parsed_op = int(op)
    except ValueError:
        parsed_op = op
    finally:
        return parsed_op

# --- TEST SUITE ---

tests_text = r'''
1. is_valid_equation("2 + 2 = 4") should return True.
2. is_valid_equation("2 + 3 - 1 = 4") should return True.
3. is_valid_equation("8 / 2 = 4") should return True.
4. is_valid_equation("10 * 5 = 50") should return True.
5. is_valid_equation("2 - 2 = 0") should return True.
6. is_valid_equation("2 + 9 / 3 = 5") should return True.
7. is_valid_equation("20 - 2 * 3 = 14") should return True.
8. is_valid_equation("2 + 5 = 6") should return False.
9. is_valid_equation("10 - 2 * 3 = 24") should return False.
10. is_valid_equation("3 + 9 / 3 = 4") should return False.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)